File: skills/ayran/src/ayran/gates/credentials.py
"""Short-TTL, identity-bound, single-use credentials (spec §11.4).

Key GENERATION lives only in the Prime extension (TypeScript, outside
model-reachable Python): the sidecar constructs its
:class:`CredentialAuthority` from the key enrolled via the
``credentials.enroll`` RPC and then only VERIFIES and CONSUMES tokens against
it. Per-spawn challenger tokens are minted extension-side and vaulted
server-side via ``credentials.deliver``; the model never handles a challenger
token. Honest HMAC caveat: the key is symmetric, so post-enrollment the
sidecar COULD mint; mint-separation for challenger tokens is procedural (the
sidecar mints only the §11.4-sanctioned session-writer derivations), while
single-use / TTL / child-binding enforcement remains cryptographic and
server-side. ``verify(consume=True)`` enforces the exactly-one-submission
grant and is called at the verdict seal; the token is dead afterwards (reuse
is refused). ``revoke`` retires a token early.

Wire format (byte-compatible with ``prime/extension/credentials.ts``):
``body = base64url(JSON.stringify(payload with SORTED keys), unpadded)`` and
``signature = HMAC-SHA256 lowercase hex over body``; token = ``body + "." +
signature``. Payload fields are exactly ``jti`` (16 hex), ``iat``, ``exp``,
``grants`` (sorted), ``writer``, ``child_id``, ``session``.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import secrets
import time
from collections.abc import Callable, Mapping
from typing import Any

CREDENTIAL_GRANT_REMEMBER = "hypotheses.remember"
CREDENTIAL_GRANT_GATE_A = "gate_a_submission"
DEFAULT_TTL_SECONDS = 300.0

CLOCK = Callable[[], float]


class CredentialError(ValueError):
    """Stable refusal reason for a challenger/session credential."""

    def __init__(self, reason: str, details: dict[str, Any] | None = None) -> None:
        super().__init__(reason)
        self.reason = reason
        self.details = details or {}

    def as_dict(self) -> dict[str, Any]:
        value: dict[str, Any] = {"code": f"CREDENTIAL_{self.reason}", "reason": self.reason}
        if self.details:
            value["details"] = self.details
        return value


def _b64encode(payload: bytes) -> str:
    return base64.urlsafe_b64encode(payload).decode("ascii").rstrip("=")


def _b64decode(body: str) -> bytes:
    padding = "=" * (-len(body) % 4)
    return base64.urlsafe_b64decode(body + padding)


class CredentialAuthority:
    """Short-TTL, identity-bound, single-use credentials (§11.4).

    Tokens are HMAC-SHA256 signed compact JSON. ``verify(consume=True)`` enforces
    the exactly-one-submission grant and is called at the verdict seal; the token
    is dead afterwards (reuse is refused). ``revoke`` retires a token early.
    """

    def __init__(
        self,
        key: bytes | None = None,
        *,
        ttl_seconds: float = DEFAULT_TTL_SECONDS,
        clock: CLOCK | None = None,
    ) -> None:
        self._key = key or secrets.token_bytes(32)
        self.ttl_seconds = float(ttl_seconds)
        self._clock: CLOCK = clock or time.time
        self._used: set[str] = set()
        self._revoked: set[str] = set()

    def mint(
        self,
        *,
        grants: tuple[str, ...] | list[str],
        writer: Mapping[str, str] | None = None,
        child_id: str | None = None,
        session: str | None = None,
        ttl_seconds: float | None = None,
    ) -> str:
        now = self._clock()
        lifetime = self.ttl_seconds if ttl_seconds is None else float(ttl_seconds)
        payload = {
            "jti": secrets.token_hex(8),
            "iat": now,
            "exp": now + lifetime,
            "grants": sorted(str(grant) for grant in grants),
            "writer": dict(writer) if writer else None,
            "child_id": str(child_id) if child_id else None,
            "session": str(session) if session else None,
        }
        body = _b64encode(
            json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")
        )
        signature = hmac.new(self._key, body.encode("ascii"), hashlib.sha256).hexdigest()
        return f"{body}.{signature}"

    def verify(
        self,
        token: str,
        *,
        grant: str | None = None,
        child_id: str | None = None,
        consume: bool = False,
    ) -> dict[str, Any]:
        if not isinstance(token, str) or token.count(".") != 1:
            raise CredentialError("MALFORMED", details={"token_type": type(token).__name__})
        body, signature = token.split(".", 1)
        expected = hmac.new(self._key, body.encode("ascii"), hashlib.sha256).hexdigest()
        if not hmac.compare_digest(signature, expected):
            raise CredentialError("SIGNATURE_INVALID")
        try:
            payload = json.loads(_b64decode(body))
        except (ValueError, UnicodeError) as error:
            raise CredentialError("MALFORMED", details={"parse": str(error)}) from error
        if not isinstance(payload, dict):
            raise CredentialError("MALFORMED")
        jti = str(payload.get("jti") or "")
        if not jti:
            raise CredentialError("MALFORMED")
        if jti in self._revoked:
            raise CredentialError("REVOKED", details={"jti": jti})
        if self._clock() >= float(payload.get("exp") or 0.0):
            raise CredentialError("EXPIRED", details={"jti": jti})
        if jti in self._used:
            # Single-use grant: reuse after the verdict seal is refused (§11.4).
            raise CredentialError("REVOKED", details={"jti": jti, "reuse": True})
        grants = {str(item) for item in (payload.get("grants") or [])}
        if grant is not None and grant not in grants:
            raise CredentialError("GRANT_MISMATCH", details={"required": grant, "grants": sorted(grants)})
        if child_id is not None and str(payload.get("child_id") or "") != child_id:
            raise CredentialError("BOUND_ID_MISMATCH", details={"required": child_id})
        if consume:
            self._used.add(jti)
        return payload

    def revoke(self, token: str) -> None:
        try:
            body = token.split(".", 1)[0]
            payload = json.loads(_b64decode(body))
            jti = str(payload.get("jti") or "")
        except (IndexError, ValueError, UnicodeError):
            return
        if jti:
            self._revoked.add(jti)

    def consume(self, token: str) -> None:
        """Retire the token's single grant — called at the verdict seal (§11.4)."""

        payload = self.verify(token)
        jti = str(payload.get("jti") or "")
        if jti:
            self._used.add(jti)

    def mint_session_writer(
        self,
        *,
        session: str,
        writer_kind: str,
        writer_id: str,
        ttl_seconds: float | None = None,
    ) -> str:
        return self.mint(
            grants=(CREDENTIAL_GRANT_REMEMBER,),
            writer={"kind": writer_kind, "id": writer_id},
            session=session,
            ttl_seconds=ttl_seconds,
        )

    def mint_challenger(
        self,
        *,
        child_id: str,
        ttl_seconds: float | None = None,
    ) -> str:
        return self.mint(
            grants=(CREDENTIAL_GRANT_GATE_A,),
            child_id=child_id,
            ttl_seconds=ttl_seconds,
        )

    def state(self) -> dict[str, Any]:
        return {
            "ttl_seconds": self.ttl_seconds,
            "used_credentials": len(self._used),
            "revoked_credentials": len(self._revoked),
        }

File: prime/extension/credentials.ts
/**
 * Extension-side credential minting (spec §11.4, R5 W7 relocation).
 *
 * Key GENERATION lives only here — outside model-reachable Python. The sidecar
 * receives the key once via the `credentials.enroll` RPC (bearer-authenticated
 * channel; the key never appears in env values, argv, or logs) and afterwards
 * only verifies and consumes tokens against it. Per-spawn challenger tokens
 * are minted here and vaulted server-side via `credentials.deliver`; the model
 * never handles a challenger token.
 *
 * Wire format — byte-compatible with `ayran.gates.credentials` (Python):
 *   body       = base64url(JSON.stringify(payload with SORTED keys), unpadded)
 *   signature  = HMAC-SHA256 lowercase hex over body (ascii)
 *   token      = body + "." + signature
 * Payload fields are exactly jti (16 hex), iat, exp, grants (sorted), writer,
 * child_id, session. Honest HMAC caveat: the key is symmetric, so
 * post-enrollment the sidecar COULD mint; mint-separation for challenger
 * tokens is procedural, while single-use / TTL / child-binding enforcement
 * stays cryptographic and server-side.
 */

import { createHmac, randomBytes } from "node:crypto";
import type { RuntimeState } from "./bootstrap.ts";

export const CREDENTIAL_GRANT_REMEMBER = "hypotheses.remember";
export const CREDENTIAL_GRANT_GATE_A = "gate_a_submission";
export const DEFAULT_TTL_SECONDS = 300;
export const SESSION_KEY_BYTES = 32;

export type CredentialGrant =
  | typeof CREDENTIAL_GRANT_REMEMBER
  | typeof CREDENTIAL_GRANT_GATE_A;

export interface CredentialPayload {
  jti: string;
  iat: number;
  exp: number;
  grants: string[];
  writer: Record<string, string> | null;
  child_id: string | null;
  session: string | null;
}

export interface MintOptions {
  grants: CredentialGrant[];
  writer?: Record<string, string>;
  child_id?: string;
  session?: string;
  ttlSeconds?: number;
  nowSeconds?: number;
  jti?: string;
}

/** Deep key-sorted JSON.stringify (Python json.dumps(sort_keys=True) parity). */
function sortedStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value) as string;
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => sortedStringify(item)).join(",")}]`;
  }
  const record = value as Record<string, unknown>;
  const keys = Object.keys(record).sort();
  return `{${keys
    .map((key) => `${JSON.stringify(key)}:${sortedStringify(record[key])}`)
    .join(",")}}`;
}

export function encodeCredentialBody(
  payload: CredentialPayload,
): string {
  return Buffer.from(sortedStringify(payload), "utf8").toString("base64url");
}

export function signCredentialBody(
  body: string,
  key: Buffer,
): string {
  return createHmac("sha256", key).update(body, "ascii").digest("hex");
}

export class CredentialMinter {
  /** 32-byte random session key; generated here, enrolled once, never logged. */
  readonly key: Buffer;

  constructor(key: Buffer = randomBytes(SESSION_KEY_BYTES)) {
    if (key.length !== SESSION_KEY_BYTES) {
      throw new Error(`credential session key must be ${SESSION_KEY_BYTES} bytes`);
    }
    this.key = key;
  }

  keyB64(): string {
    return this.key.toString("base64");
  }

  mint(options: MintOptions): string {
    const iat = options.nowSeconds ?? Math.floor(Date.now() / 1000);
    const ttl = options.ttlSeconds ?? DEFAULT_TTL_SECONDS;
    const payload: CredentialPayload = {
      jti: options.jti ?? randomBytes(8).toString("hex"),
      iat,
      exp: iat + ttl,
      grants: [...options.grants].sort(),
      writer: options.writer ?? null,
      child_id: options.child_id ?? null,
      session: options.session ?? null,
    };
    const body = encodeCredentialBody(payload);
    return `${body}.${signCredentialBody(body, this.key)}`;
  }

  mintChallenger(childId: string, ttlSeconds = DEFAULT_TTL_SECONDS): string {
    return this.mint({
      grants: [CREDENTIAL_GRANT_GATE_A],
      child_id: childId,
      ttlSeconds,
    });
  }
}

export interface EnrollOutcome {
  attempted: boolean;
  accepted: boolean;
  code?: string;
}

/**
 * Enroll the runtime's session key with the sidecar. Idempotent per runtime:
 * once an enrollment is accepted it is never retried (the sidecar refuses a
 * second enrollment with CREDENTIAL_ALREADY_ENROLLED, fail-closed).
 */
export async function enrollCredentials(
  runtime: RuntimeState,
): Promise<EnrollOutcome> {
  if (!runtime.sidecar || runtime.credentialsEnrolled) {
    return { attempted: false, accepted: runtime.credentialsEnrolled };
  }
  const result = (await runtime.sidecar.tryCall("credentials.enroll", {
    key_b64: runtime.credentials.keyB64(),
  })) as { accepted?: boolean; error?: { code?: string } } | undefined;
  const accepted = Boolean(result?.accepted);
  const code = result?.error?.code;
  if (accepted) {
    runtime.credentialsEnrolled = true;
    runtime.telemetry.event("info", "ayran.credentials.enrolled", {
      outcome: "accepted",
    });
  } else {
    runtime.telemetry.event("warn", "ayran.credentials.enroll_refused", {
      outcome: "degraded",
      error_code: code ?? "sidecar_unreachable",
    });
  }
  return { attempted: true, accepted, code };
}

/**
 * Mint a child-bound gate_a_submission token and vault it server-side
 * (§11.4/W7). Failures are tolerated silently: child tracking and degraded
 * operation must never depend on credential delivery, and tokens/key material
 * are never logged.
 */
export async function deliverChildCredential(
  runtime: RuntimeState,
  childId: string,
): Promise<void> {
  if (!runtime.sidecar || !childId) {
    return;
  }
  try {
    const token = runtime.credentials.mintChallenger(childId);
    await runtime.sidecar.tryCall("credentials.deliver", {
      child_id: childId,
      // "credential", never "token": params.token is the run bearer on the wire.
      credential: token,
    });
  } catch {
    // Best-effort: an undeliverable vault entry only means Gate A must use an
    // explicit credential channel later.
  }
}
